js-packages/@prestojs/rest/src/Endpoint.ts

```diff
@@ -127,6 +127,38 @@ type MiddlewareContext<T> = {
      * The original options passed to [execute](doc:Endpoint#method-execute)
      */
     executeOptions: EndpointExecuteOptions;
+    /**
+     * The endpoint the middleware is executing within.
+     */
+    endpoint: Endpoint;
+    /**
+     * The resolved URL.
+     *
+     * Only available in the response part of the middleware.
```

Comment:
no because you only have one type in the middleware, eg.

```
function myMiddleware(urlConfig, requestInit, next, context) {
   // currently url not available
  await next(urlConfig, requestInit);
  // now it is but `context` is still the same
}
```

Reply:
Yes, you would end up with 2 variables:

```js
function myMiddleware(next, urlConfig, requestInit, requestContext) {
   // currently url not available
  responseContext = await next(urlConfig, requestInit, requestContext);
  // now it is but `context` is still the same
}
```

(have reordered the order of arguments to make it a little clearer)

Reply:
I don't particularly like `next` returning context instead of response. The middleware returns the response (and it should... too cumbersome and error prone to expect it to also return context) so seems confusing for `next` to return context instead.

I'm inclined to leave it with runtime checking given changing it would really only give you slightly better types at the cost of a more complicated interface

Reply:
It is nice having the URL and response etc only available after the call to next but not sure what the return should be from `next`. It needs to be possible just to return direct so we'd have to support that use case, eg.

```js
function requestOnlyMiddleware(urlConfig, requestInit, next, context) {
   // do something to requestInit
   return next(urlConfig, requestInit);
}
```

If next was to return the relevant context for response that would give access to `url`, `response` and `decodedBody`. But most of the time you just want to do something to the transformed decodedBody from the previous middleware... so how should that be accessed? Another key, eg `result`?

```js
function middleware(urlConfig, requestInit, next) {
  const { decodedBody, url, response, result } = await next(urlConfig, requestInit);

  return transformAndCache(result);
}
```

or array return

```js
function middleware(urlConfig, requestInit, next) {
  const [result, { decodedBody, url, response }] = await next(urlConfig, requestInit);
  // Or if you don't need any of the response context:
  //   const [result] = await next(urlConfig, requestInit);

  return transformAndCache(result);
}
```

Any thoughts?

Do we want to change the order of arguments to have `next` first?

Reply:
Yes please to changing order of arguments.

To clarify what you're saying: the return value from `next` will be the result + the fields that were previously only available to MiddlewareContext after the request?

For fields that were available prior to that you'd just use the same MiddlewareContext you had before -- there's nothing the other middleware could have changed that we'd care about? (eg executeOptions?)

re the syntax, I prefer the first one. From the perspective of a user I don't see any difference between `result` and `url` -- they're still all just the output of the next() call.

Reply:
The difference is you have to return result whereas url is there in case you need it